### ISSUE TITLE
Survey of TensorFlow-probability

### ISSUE THREAD

**Issue by daigo0927** [daigo0927/blog#3]
Survey on Bayesian neural network with probabilistic layers (reparameterization and flipout).

**Comment by daigo0927** [comment]
Experiment plan: implement simple CNNs of vanilla one, the one using the re-parameterization trick, and using the Flipout technique.
Compare learning dynamics (i.e. elapsed time, posterior distribution) and accuracy.
https://www.tensorflow.org/probability/api_docs/python/tfp/layers/Convolution2DFlipout

**Comment by daigo0927** [comment]
**Warning**; Construct a model in `@tf.function` decorated function causes unexistence of KL-div regularization term (exists in the official example: https://www.tensorflow.org/probability/api_docs/python/tfp/layers/Convolution2DFlipout).
This unexistence is a serious problem in Bayesian learning. I would solve this by initializing the model with dummy inputs.

**Comment by daigo0927** [comment]
**Warning**: Dummy inference creates `model.losses` variables which are described above.
However, after executing `train_step`, `model.losses` lost. This behavior happens regardless of decoration of `tf.function`.

**Comment by daigo0927** [comment]
This issue is difficult to solve for me with `tf.GradientTape` API.
I would follow the official example: https://github.com/tensorflow/probability/blob/master/tensorflow_probability/examples/bayesian_neural_network.py which uses Keras API
